### ISSUE TITLE
Getting @typeInfo of a struct inside that struct causes Zig to segfault

### ISSUE THREAD

**Issue by heidezomp** [ziglang/zig#2968]
I was exploring the boundaries of the compile-time reflection capabilities of Zig (which are very cool!!) when I found that calling `@typeInfo` on a struct inside that same struct causes the compiler to segfault. Doing this probably isn't useful; I can't think of what the result should be (is this is even theoretically possible?). So this should probably result in a compile-time error?

Example code: 
```zig
const foo = struct {
    const info = @typeInfo(foo);
};

pub fn main() void {
    const info = foo.info;
}
```
```
% zig build-exe typeinfo_segfault.zig
Segmentation fault at address 0x18
???:?:?: 0x557480bade39 in ??? (???)


zsh: segmentation fault  zig build-exe typeinfo_segfault.zig
```
Backtrace in gdb:
```
(gdb) run
Starting program: /home/heidezomp/zig/build/bin/zig build-exe typeinfo_segfault.zig
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x0000555555a61e39 in ir_make_type_info_decls(IrAnalyze*, IrInstruction*, ConstExprValue*, ScopeDecls*) ()
(gdb) bt
#0  0x0000555555a61e39 in ir_make_type_info_decls(IrAnalyze*, IrInstruction*, ConstExprValue*, ScopeDecls*) ()
#1  0x0000555555a62dd0 in ir_make_type_info_value(IrAnalyze*, IrInstruction*, ZigType*, ConstExprValue**) ()
#2  0x0000555555aa32d9 in ir_analyze_instruction_base(IrAnalyze*, IrInstruction*) ()
#3  0x0000555555aa418b in ir_analyze(CodeGen*, IrExecutable*, IrExecutable*, ZigType*, AstNode*) ()
#4  0x0000555555aa448e in ir_eval_const_value(CodeGen*, Scope*, AstNode*, ZigType*, unsigned long*, unsigned long*, ZigFn*, Buf*, AstNode*, Buf*, IrExecutable*, AstNode*) ()
#5  0x0000555555ad3c7f in analyze_const_value(CodeGen*, Scope*, AstNode*, ZigType*, Buf*) ()
#6  0x0000555555ae56e4 in resolve_decl_var(CodeGen*, TldVar*) ()
#7  0x0000555555ae7a2b in resolve_top_level_decl(CodeGen*, Tld*, AstNode*) [clone .part.76] ()
#8  0x0000555555a8b724 in ir_analyze_decl_ref(IrAnalyze*, IrInstruction*, Tld*) ()
#9  0x0000555555a9ee10 in ir_analyze_instruction_field_ptr(IrAnalyze*, IrInstructionFieldPtr*) ()
#10 0x0000555555aa2c93 in ir_analyze_instruction_base(IrAnalyze*, IrInstruction*) ()
#11 0x0000555555aa418b in ir_analyze(CodeGen*, IrExecutable*, IrExecutable*, ZigType*, AstNode*) ()
#12 0x0000555555ae0df9 in analyze_fn_ir(CodeGen*, ZigFn*, AstNode*) ()
#13 0x0000555555ae0b1f in analyze_fn_body(CodeGen*, ZigFn*) [clone .part.77] ()
#14 0x0000555555ae7ea3 in semantic_analyze(CodeGen*) ()
#15 0x0000555555a4df68 in codegen_build_and_link(CodeGen*) ()
#16 0x0000555555a1c0c5 in main ()
```
This is using a release build of Zig from the master branch.

**Comment by emekoi** [comment]
I think this will be solved by #2174.

**Comment by heidezomp** [comment]
@emekoi Thanks for the pointer, that looks like an interesting issue! I will try to digest it as best I can, and hopefully learn more about how Zig works internally along the way :smile:

**Comment by heidezomp** [comment]
This was not solved by #2174.

**Comment by andrewrk** [comment]
This now gives:

```
./test2.zig:2:18: error: dependency loop detected
    const info = @typeInfo(foo);
                 ^
./test2.zig:6:21: note: referenced here
    const info = foo.info;
                    ^
```

Related: #3893
